typing/typecore.ml

```diff
@@ -3025,12 +3098,46 @@ and type_expect_
         if rec_flag = Recursive then In_rec
         else if List.compare_length_with spat_sexp_list 1 > 0 then In_group
         else With_attributes in
-      let (pat_exp_list, new_env, unpacks) =
-        type_let existential_context env rec_flag spat_sexp_list true in
-      let body = type_unpacks new_env unpacks sbody ty_expected_explained in
-      let () =
-        if rec_flag = Recursive then
-          check_recursive_bindings env pat_exp_list
+      let may_contain_modules =
+        List.exists (fun pvb -> may_contain_modules pvb.pvb_pat) spat_sexp_list
+      in
+      let (pat_exp_list, body, _new_env) =
+        with_local_level_if may_contain_modules begin fun () ->
+          let allow_modules =
+            if may_contain_modules
+            then
+              let scope = create_scope () in
+              Modules_allowed { scope }
+            else Modules_rejected
+          in
+          let (pat_exp_list, new_env) =
+            type_let existential_context env rec_flag spat_sexp_list
+              allow_modules
+          in
+          let body = type_expect new_env sbody ty_expected_explained in
+          let () =
+            if rec_flag = Recursive then
+              check_recursive_bindings env pat_exp_list
+          in
+          (pat_exp_list, body, new_env)
+        end
+        (* If the patterns contain module unpacks, there is a possibility that
+           the types of the let body or let definitions mention types
+           introduced by those unpacks. (The latter can only happen with
+           recursive definitions.) Here, we check for scope escape via both of
+           these pathways (body, definitions).
+
+           Checking unification within an environment extended with the module
+           bindings allows us to correctly accept more programs. This
+           environment allows unification to identify more cases where a type
+           introduced by the module is equal to a type introduced at an outer
+           scope. *)
+        ~post:(fun (pat_exp_list, body, new_env) ->
+          unify_exp new_env body (newvar ());
+          if rec_flag = Recursive then
+            List.iter
+              (fun vb -> unify_exp new_env vb.vb_expr (newvar ()))
```

Comment:
Noted, thanks. I am going to double-check a detail of my updated approach with my London colleagues before pushing the updated implementation for your review.

Reply:
I've implemented the approach you suggested and it seems like it works (now pushed) [1]. But, taking a step back, the code and reasoning behind it is rather complex for a feature that probably is not widely used, namely recursive first-class module unpacks:

```ocaml
let rec (module M) = e1 in e2
```

I agree that your suggestion is better than the approach I originally took. But, in discussion with @lpw25 , we wondered if it would be better to instead disallow recursive unpacks. This lowers the maintenance burden of this PR (we get to delete [1]). Leo also points out that it's a bit odd that unpacks are allowed as `P` in `let rec P = ...` when basically nothing else is.

I think we should probably not block this PR on the decision of whether to remove recursive unpacks. Depending on what you think, we could explore making another PR to do this.

[1]
```ocaml
          if rec_flag = Recursive && may_contain_modules then begin
            List.iter
              (fun vb ->
                 (* [type_let] already generalized bound expressions' types
                    in-place. We first take an instance before checking scope
                    escape at the outer level to avoid losing generality of
                    types added to [new_env].
                 *)
                let bound_exp = vb.vb_expr in
                generalize_structure_exp bound_exp;
                let bound_exp_type = Ctype.instance bound_exp.exp_type in
                let loc = proper_exp_loc bound_exp in
                let outer_var = newvar2 outer_level in
                unify_exp_types loc new_env bound_exp_type outer_var)
              pat_exp_list
          end;
```

Reply:
I don't think unpacks in `let rec` patterns are currently allowed:
```ocaml
# let rec (module M) = (module struct end);;
Error: Modules are not allowed in this pattern.
```
I get the same result in both 4.14 and 5.0 (haven't tried trunk, but I don't remember any change related to that). I think we'd all rather keep them forbidden, as `let rec` is already complicated enough as it is.

Reply:
They're allowed for `let .. in`:
```ocaml
# let rec (module M : S) = (module struct end) in ();;
- : unit = ()
```
but I really think they shouldn't be.

Reply:
I agree that maintenance is a burden, and that I don't see much practical use for them.
One can always use the old approach expanded as syntactic sugar anyway.

However, if we do not allow them we need a good rationale for it. Otherwise this is yet another irregularity in the language.
Since I don't see one myself (the only restrictions on recursion are supposed to be about initialization, and this is detected independently), I would tend to keep the support for the sake of regularity. The point is that the feature is not "recursive module unpack" but "module pattern", which is orthogonal to recursion.

BTW, there is a reasonably good rationale for disallowing module patterns in toplevel let definitions, but this still confuses people.

Reply:
> (the only restrictions on recursion are supposed to be about initialization, and this is detected independently)

No, there is also a restriction that the only allowed patterns for recursive bindings are variables. That's why I was surprised that module patterns were allowed.

And the feature in its current state is very limited: if you have to insert a coercion between the actual structure and its declared signature, then all definitions that are actually recursive are forbidden (I suspect that the coercion makes `Rec_check` classify the definition as `Dynamic` instead of `Static`, but I haven't tried to print the intermediate typedtree to see why). So you don't have access to more features than just unpacking after the definition:
```ocaml
(* This is allowed, including at toplevel: *)
let rec m =
  let module M = (val m : S) in
  (module struct
    let x = 0
    let y () =
      M.x
  end: S)
module M : S' = (val m : S)
```
I'll ping @gasche, since he may have some more insights to share.

Reply:
It is indeed the case that we currently only allow variables as the pattern of a `let rec`. I think that the fact that module patterns were accepted before is an unintended side-effect of their desugaring implementation. I see no problem with rejecting them for this reason, in fact this is arguably the more regular thing to do.

(Note that if we had a `Tpat_unpack` construct as a proper pattern-former, then they would be rejected naturally by the `Illegal_letrec_pat` check.)